### ISSUE TITLE
Training my own data

### ISSUE THREAD

**Issue by shenao1994** [Ruyi-Zha/r2_gaussian#8]
I combine the [DiffDRR ](https://github.com/eigenvivek/DiffDRR) package to generate DRR images. I rewrote the code for generating data. However, after training, the reconstruction and rendering effects are relatively poor. I suspect it's a problem with the CT coordinate system. I referred to your process_raw_data.exe code. I found that some data needs to be processed with 'xy_invert' or 'z_invert'. What is the basis for this decision? In addition, the drr image I generated is shown in the following figure. 
![test](https://github.com/user-attachments/assets/b28509f3-7cb0-4d83-93b6-d8222094f0f6)

The original DRR image generated from the first image is normalized to 0 to 1. The second image is the GT DRR image output after running test.by I don't know what processing was done. The following image shows the reconstructed CT.
![recon_test](https://github.com/user-attachments/assets/fe474798-6cf3-4172-ae16-bffd4774ce5e)

The code for generating pickle format is as follows Can you help me see what processing is missing?

```
import matplotlib.pyplot as plt
import torch
from diffdrr.drr import DRR
from diffdrr.data import read
import numpy as np
from diffdrr.pose import convert
import cv2
import os
import pickle
import SimpleITK as sitk
from data_generator.synthetic_dataset.process_raw_data import reshape_vol
def generate_data():
    ct_path = 'data/resampled_vert/dukemei.nii.gz'
    outputPath = 'data/processed/{}_vert.pickle'.format(os.path.split(ct_path)[-1].split('.')[0])
    ct_img = sitk.ReadImage(ct_path)
    raw_data = rotate_volume(ct_img)
    # raw_data = sitk.GetArrayFromImage(ct_img)
    raw_data = raw_data.transpose([2, 1, 0])
    # data_min = raw_data.min()
    # data_max = raw_data.max()
    # raw_data = (raw_data - data_min) / (data_max - data_min)
    # raw_data = raw_data.clip(0.0, 1.0)
    # print(raw_data.shape)
    size = ct_img.GetSize()
    # print(size)
    spacing = ct_img.GetSpacing()
    print(spacing)
    subject = read(ct_path, bone_attenuation_multiplier=10.5)
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    SDD = 1100.0
    Xray_H = 256
    DELX = 1.5
    # print(DELX)
    drr = DRR(
        subject,  # An object storing the CT volume, origin, and voxel spacing
        sdd=SDD,  # Source-to-detector distance (i.e., focal length)
        height=Xray_H,  # Image height (if width is not provided, the generated DRR is square)
        delx=DELX,  # Pixel spacing (in mm)
        reverse_x_axis=False
    ).to(device)
    true_params = {
        "sdr": SDD,
        "alpha": 0.0,
        "beta": 0.0,
        "gamma": 0.0,
        "bx": 0.0,
        "by": 500.0,
        "bz": 0.0,
    }
    data = {'numTrain': 50, 'numVal': 50, 'DSD': SDD, 'DSO': true_params["by"], 'nDetector': [Xray_H, Xray_H],
            'dDetector': [DELX, DELX], 'nVoxel': [size[0], size[1], size[2]],
            'dVoxel': [spacing[0], spacing[1], spacing[2]], 'offOrigin': [0.0, 0.0, 0.0],
            'offDetector': [0, 0], 'accuracy': 0.5, 'mode': 'cone', 'filter': None, 'totalAngle': 180.0,
            'startAngle': 0.0, 'randomAngle': False, 'convert': True, 'rescale_slope': 1.0, 'rescale_intercept': 0,
            'normalize': True, 'noise': 0}
    train_angles = np.linspace(0, data["totalAngle"] / 180 * np.pi, data["numTrain"] + 1)[:-1] + data[
        "startAngle"] / 180 * np.pi
    data["train"] = {"angles": train_angles}
    print(data["train"])
    data["image"] = raw_data.astype(np.float32)
    val_angles = np.sort(np.random.rand(data["numVal"]) * 180 / 180 * np.pi) + data["startAngle"] / 180 * np.pi
    data["val"] = {"angles": val_angles}
    train_projections = []
    count = 0
    drr_save_fold = 'data/drr_data'
    for rotX in train_angles:
        count += 1
        rotations, translations, pose = get_initial_parameters(true_params, rotX)
        img = drr(pose)
        proj_img = img.squeeze(0).cpu().numpy()
        proj_img = (proj_img - proj_img.min()) / (proj_img.max()-proj_img.min())
        normalized_img = cv2.normalize(img.squeeze().cpu().numpy(), None, alpha=0, beta=1,
                                       norm_type=cv2.NORM_MINMAX, dtype=cv2.CV_32F)
        normalized_img = (normalized_img * 255).astype(np.uint8)
        rgb_image = cv2.cvtColor(normalized_img, cv2.COLOR_GRAY2RGB)
        cv2.imwrite(os.path.join(drr_save_fold, str(count).zfill(6) + '.png'), rgb_image)
        train_projections.append(proj_img)
    train_projections = np.concatenate(train_projections, axis=0)
    data["train"]["projections"] = train_projections
    val_projections = []
    for rotZ in val_angles:
        rotations, translations, pose = get_initial_parameters(true_params, rotZ)
        img = drr(pose)
        proj_img = img.squeeze(0).cpu().numpy()
        proj_img = (proj_img - proj_img.min()) / (proj_img.max()-proj_img.min())
        normalized_img = cv2.normalize(img.squeeze().cpu().numpy(), None, alpha=0, beta=1, norm_type=cv2.NORM_MINMAX,
                                       dtype=cv2.CV_32F)
        normalized_img = (normalized_img * 255).astype(np.uint8)
        rgb_image = cv2.cvtColor(normalized_img, cv2.COLOR_GRAY2RGB)
        cv2.imwrite(os.path.join(drr_save_fold, str(count).zfill(6) + '.png'), rgb_image)
        val_projections.append(proj_img)
    val_projections = np.concatenate(val_projections, axis=0)
    data["val"]["projections"] = val_projections
    print(train_angles.shape)
    print(train_projections.shape)
    with open(outputPath, "wb") as handle:
        pickle.dump(data, handle, pickle.HIGHEST_PROTOCOL)


def pose_from_carm(sid, tx, ty, alpha, beta, gamma):
    rot = torch.tensor([[alpha, beta, gamma]], dtype=torch.float32)
    xyz = torch.tensor([[tx, sid, ty]], dtype=torch.float32)
    return convert(rot, xyz, parameterization="euler_angles", convention="ZXY")


def get_initial_parameters(true_params, deltaZ):
    alpha = true_params["alpha"] + deltaZ
    beta = true_params["beta"]
    gamma = true_params["gamma"]
    bx = true_params["bx"]
    by = true_params["by"]
    bz = true_params["bz"]
    pose = pose_from_carm(by, bx, bz, alpha, beta, gamma).cuda()
    rotations, translations = pose.convert("euler_angles", "ZXY")
    return rotations, translations, pose


def rotate_volume(ct_image):
    angle_degrees = -90 
    angle_radians = angle_degrees * (np.pi / 180.0)

    size = ct_image.GetSize()
    spacing = ct_image.GetSpacing()
    origin = ct_image.GetOrigin()

    center_physical = [
        origin[i] + 0.5 * (size[i] - 1) * spacing[i] for i in range(3)
    ]

    rotation_center = center_physical
    rotation_transform = sitk.Euler3DTransform()
    rotation_transform.SetCenter(rotation_center)
    rotation_transform.SetRotation(0.0, 0.0, angle_radians)

    interpolator = sitk.sitkLinear  
    output_pixel_type = ct_image.GetPixelID()
    default_value = 0  

    rotated_image = sitk.Resample(
        ct_image,
        ct_image.GetSize(),
        rotation_transform,
        interpolator,
        ct_image.GetOrigin(),
        ct_image.GetSpacing(),
        ct_image.GetDirection(),
        default_value,
        output_pixel_type
    )

    ct_array = sitk.GetArrayFromImage(rotated_image)
    # Normalize CT image to [0, 1]
    ct_array = (ct_array - ct_array.min()) / (ct_array.max() - ct_array.min())
    return ct_array


if __name__ == '__main__':
    generate_data()
```

**Comment by Ruyi-Zha** [comment]
Hi,

Some raw CT volumes may have unusual orientations, such as being upside-down. To correct this, we use 'xy_invert' and 'z_invert' to align them to a standard orientation. This adjustment shouldn’t impact rendering or reconstruction.

I haven’t used DiffDRR before; I use TIGRE to generate DRRs. I recommend first checking if DRRs using two packages match under the same scanner settings.

 If you're generating a dataset in pickle format, ensure that the units in the pickle files align with those used in SAX-NeRF. I recall that SAX-NeRF’s geometry parameters are in mm, while the X-ray projections are in meters. You may need to manually scale the geometry parameters to meters like [this](https://github.com/Ruyi-Zha/r2_gaussian/blob/75bb8a91439cbf45a17b4083ef8a0552ec52b5a7/r2_gaussian/dataset/dataset_readers.py#L200).

If you still encounter issues, feel free to share your data with me privately, and I can help review it to identify the problem.

**Comment by shenao1994** [comment]
Thank you for your patient answer! The first picture is the training data generated by **DiffDRR**.  GT is the DRR image generated at 0 degrees. But the gt.png (second column in the image) obtained from the output of test. py becomes very strange. It should be the same image representation as GT. It makes me feel strange. 

When I trained X-Gaussian using the data I generated myself, the results were normal. I will first check some of the issues you mentioned myself. 

 Can you give me your email? If there are any further questions, I will transfer the data to you. Thank you again for your help!

**Comment by Ruyi-Zha** [comment]
It's ruyi.zha@gmail.com.
